src/spurt/workflows/emcf/_solver.py

```diff
@@ -0,0 +1,266 @@
+        """Number of points in the spatial network."""
+        return self._solver_space.npoints
+
+    @property
+    def nlinks(self) -> int:
+        """Number of links in the spatial network."""
+        return self._solver_space.nedges
+
+    @property
+    def nepochs(self) -> int:
+        """Number of points in the temporal network."""
+        return self._solver_time.npoints
+
+    @property
+    def nifgs(self) -> int:
+        """Number of links in the temporal network."""
+        return self._solver_time.nedges
+
+    @property
+    def settings(self) -> Settings:
+        """Retrieve settings for the workflow."""
+        return self._settings
+
+    @property
+    def link_model(self) -> LinkModelInterface | None:
+        """Retrieve the link model for the workflow."""
+        return self._link_model
+
+    def unwrap_cube(self, wrap_data: np.ndarray) -> np.ndarray:
+        """Unwrap a 3D cube of data.
+
+        Parameters
+        ----------
+        wrap_data: np.ndarray
+            2D array of size (nslc, npoints) or (nifg, npoints).
+
+        Returns
+        -------
+        uw_data: np.ndarray
+            2D float32 array of size (nifg, npoints).
+        """
+        if wrap_data.ndim != 2:
+            errmsg = f"Input data is not a 2D array - {wrap_data.ndim}"
+            raise ValueError(errmsg)
+
+        if wrap_data.shape[0] == self.nepochs:
+            input_is_ifg = False
+        elif wrap_data.shape[0] == self.nifgs:
+            input_is_ifg = True
+        else:
+            errmsg = (
+                f"Input size {wrap_data.shape[0]} does not match solver"
+                f" for {self.nifgs} Ifgs from {self.nepochs} images"
+            )
+            raise ValueError(errmsg)
+
+        # First unwrap in time to get spatial gradients
+        grad_space = self.unwrap_gradients_in_time(wrap_data, input_is_ifg=input_is_ifg)
+
+        # Then unwrap spatial gradients
+        return self.unwrap_gradients_in_space(grad_space)
+
+    def unwrap_gradients_in_time(
+        self, wrap_data: np.ndarray, *, input_is_ifg: bool
+    ) -> np.ndarray:
+        """Temporally unwrap links in parallel."""
+        # First set up temporal cost
+        if self.settings.t_cost_type == "unit":
+            cost = np.ones(self.nifgs, dtype=int)
+        elif self.settings.t_cost_type == "distance":
+            cost = utils.distance_costs(
+                self._solver_time.points,
+                self._solver_time.edges,
+                scale=self.settings.t_cost_scale,
+            )
+        elif self.settings.t_cost_type == "centroid":
+            cost = utils.centroid_costs(
+                self._solver_time.points,
+                self._solver_time.cycles,
+                self._solver_time.dual_edges,
+                scale=self.settings.t_cost_scale,
+            )
+        else:
+            errmsg = f"Unknown cost type: {self.settings.t_cost_type}"
+            raise ValueError(errmsg)
```

Comment:
None of these temporal unwrapping cost functions match the actual cost function that's used in the cited paper. That's fine, but this discrepancy between the algorithm described in the paper and what's actually implemented here should be mentioned in the docstring so that we don't surprise users expecting a faithful implementation of the EMCF algorithm.

Do you have any plans to implement Pepe & Lanari's temporal unwrapping cost model as well in the future?

Reply:
None of their code is in the public domain, so naturally it is not going to be possible to replicate it as is. They also use constrained Delaunay triangulation etc - so I would say noone should expect a faithful implementation, since none of it is available in the public domain. The framework allows for coherence related costs to be applied - its a question to this team, if we are guaranteed access to these as inputs in all cases. If so, we can consider adding that as an option.